### PR TITLE
[mlr] send BMLR.ntf on Backbone link

### PR DESCRIPTION
**Pull request by simonlingoogle**
This PR enhances Backbone Router to send BMLR.ntf on the backbone link for Multicast Listeners.

- [x] Send `BMLR.ntf` to the Backbone link
- [x] Add Backbone test with packet verification
- [x] Includes some revision to the Thread 1.2 Backbone CI scripts to make it more stable 

**Receiving and processing `BMLR.ntf` on the Backbone link is not implemented in this PR**

**MLR PR List:**
- https://github.com/openthread/ot-br-posix/pull/532 add multicast forwarding for MLR Listeners
- https://github.com/openthread/openthread/pull/5388 send BMLR.ntf on Backbone link **<== This PR**

### CODE REVIEW

File: src/core/thread/thread_uri_paths.hpp [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+ * @def OT_URI_PATH_BACKBONE_MLR
+ *
+ * The URI Path for Backbone Multicast Listener Report.
+ *
+ */
+#define OT_URI_PATH_BACKBONE_MLR "b/bmr"
+
+} // namespace ot
+
+#endif // THREAD_URIS_HPP_
```
Comment (librasungirl) [inline src/core/thread/thread_uri_paths.hpp:346]: https://github.com/openthread/openthread/pull/5482 introduced src/core/thread/uri_paths.hpp for Thread URIs.

Comment (simonlingoogle) [reply src/core/thread/thread_uri_paths.hpp:346]: OMG. I added the whole file back. 👎

Comment (simonlingoogle) [reply src/core/thread/thread_uri_paths.hpp]: fixed

**Comment by simonlingoogle** [comment]
@jwhui @abtink @librasungirl 

This PR enhance PBBR to send `BMLR.ntf` on the Backbone link when a Multicast Listener is registered.
This PR also adds a test to verify the `BMLR.ntf` based on #5489 .

Now it's ready for reviewing. 

Please take a look when you have time, thanks!

File: tests/scripts/thread-cert/backbone/test_bmlr.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        self.wait_node_state(SBBR, 'router', 5)
+        self.nodes[SBBR].set_backbone_router(reg_delay=REREG_DELAY, mlr_timeout=MLR_TIMEOUT)
+        self.nodes[SBBR].enable_backbone_router()
+        self.simulator.go(5)
+        self.assertFalse(self.nodes[SBBR].is_primary_backbone_router)
+
+        self.nodes[ROUTER1].start()
+        self.wait_node_state(ROUTER1, 'router', 5)
+
+        self.nodes[PBBR].add_ipmaddr("ff04::1")
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tests/scripts/thread-cert/backbone/test_bmlr.py:95]: The MA consts in `pktverify/consts.py` is in conformance with the Test spec. But the test script here has nothing to to with certification. I will define local constants for `ff04::1` and `ff04::2`.

File: tests/scripts/thread-cert/backbone/test_bmlr.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        self.simulator.go(REREG_DELAY)
+        self.nodes[ROUTER1].add_ipmaddr("ff04::2")
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tests/scripts/thread-cert/backbone/test_bmlr.py]: I think I should add commissioner MLR test in this PR too.

[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tests/scripts/thread-cert/backbone/test_bmlr.py]: commissioner MLR test added

File: tests/scripts/thread-cert/backbone/test_bmlr.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        },
+        ROUTER1: {
+            'name': 'ROUTER1',
+            'allowlist': [PBBR, SBBR, COMMISSIONER],
+            'version': '1.2',
+            'router_selection_jitter': 1,
+        },
+        COMMISSIONER: {
+            'name': 'COMMISSIONER',
+            'allowlist': [PBBR, SBBR, ROUTER1],
```
Comment (librasungirl) [inline tests/scripts/thread-cert/backbone/test_bmlr.py:86]: should match the topology diagram
```suggestion
            'allowlist': [ROUTER1],
```

**Review by librasungirl — APPROVED** [review]

File: src/core/backbone_router/bbr_manager.cpp [hunk 5]
```diff
@@ -287,6 +299,48 @@ void Manager::SendMulticastListenerRegistrationResponse(const Coap::Message &
...
+    messageInfo.SetHopLimit(Mle::kDefaultBackboneHoplimit);
+    messageInfo.SetIsHostInterface(true);
+
+    SuccessOrExit(error = backboneTmf.SendMessage(*message, messageInfo, nullptr, nullptr));
+
+exit:
+    if (error != OT_ERROR_NONE && message != nullptr)
+    {
+        message->Free();
+    }
```
Comment (abtink) [inline src/core/backbone_router/bbr_manager.cpp:339]:
```suggestion
    FreeMessageOnError(message, error);
```

File: src/core/backbone_router/bbr_manager.cpp [hunk 6]
```diff
@@ -237,16 +239,26 @@ void Manager::HandleMulticastListenerRegistration(const Coap::Message &aMessage,
 
             if (failed)
             {
-                failedAddresses[failedAddressNum++] = address;
+                addresses[failedAddressNum++] = address;
+            }
+            else
+            {
+                // Put successfully registered addresses at the end of `addresses`.
+                addresses[kIPv6AddressesNumMax - (++successAddressNum)] = address;
```
Comment (abtink) [inline src/core/backbone_router/bbr_manager.cpp:247]: Do we need to check the success/failed array indexes to see if we reached the end of array (to ensure we never write beyond end/start of array)?

Also we write failed ones from start and success ones from end of array? Do we need to check the two indexes reach each other (or not possible)?

File: src/core/backbone_router/bbr_manager.cpp [hunk 7]
```diff
@@ -129,9 +129,11 @@ void Manager::HandleMulticastListenerRegistration(const Coap::Message &aMessage,
 
     uint16_t     addressesOffset, addressesLength;
     Ip6::Address address;
-    Ip6::Address failedAddresses[kIPv6AddressesNumMax];
-    uint8_t      failedAddressNum = 0;
-    TimeMilli    expireTime;
+    Ip6::Address addresses[kIPv6AddressesNumMax];
+    uint8_t      failedAddressNum  = 0;
+    uint8_t      successAddressNum = 0;
+
```
Comment (abtink) [inline src/core/backbone_router/bbr_manager.cpp:135]: Remove newline?

**Review by abtink — COMMENTED** [review]
Thanks @simonlingoogle. Looks good overall. 👍 
Couple of questions/suggestions below. Thanks.

File: src/core/backbone_router/bbr_manager.cpp (hunk 6 above)
Comment (simonlingoogle) [reply src/core/backbone_router/bbr_manager.cpp:247]: No. It's already guaranteed that we won't overflow `addresses` array and failed ones won't collide with success ones.

The size of `addresses` is `kIPv6AddressesNumMax`. 
We also have this check to make sure that the number of addresses in the message never exceed `kIPv6AddressesNumMax`. So accesses to `addresses` are always valid. 
```c++
    VerifyOrExit(addressesLength / sizeof(Ip6::Address) <= kIPv6AddressesNumMax,
                 status = ThreadStatusTlv::kMlrGeneralFailure);
```

Each address read from message is either `success` or `failed`. So `failedAddressNum + successAddressNum` won't exceed `kIPv6AddressesNumMax` either. So, the two indexes won't reach each other in any case.

Comment (abtink) [reply src/core/backbone_router/bbr_manager.cpp:245]: Sounds good. Thanks.

**Review by abtink — APPROVED** [review]
LGTM. Thanks. 👍

File: src/core/backbone_router/bbr_manager.cpp [hunk 8]
```diff
@@ -287,6 +297,44 @@ void Manager::SendMulticastListenerRegistrationResponse(const Coap::Message &
...
+    SuccessOrExit(error = message->InitAsNonConfirmablePost(UriPath::kBackboneMlr));
+    SuccessOrExit(error = message->SetPayloadMarker());
+
+    addressesTlv.Init();
+    addressesTlv.SetLength(sizeof(Ip6::Address) * aAddressNum);
+    SuccessOrExit(error = message->Append(&addressesTlv, sizeof(addressesTlv)));
+    SuccessOrExit(error = message->Append(aAddresses, sizeof(Ip6::Address) * aAddressNum));
+
+    SuccessOrExit(ThreadTlv::AppendUint32Tlv(*message, ThreadTlv::kTimeout, aTimeout));
```
Comment (jwhui) [inline src/core/backbone_router/bbr_manager.cpp:322]: Set error?
```suggestion
    SuccessOrExit(error = ThreadTlv::AppendUint32Tlv(*message, ThreadTlv::kTimeout, aTimeout));
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**